[Merged by Bors] - feat(linear_algebra/determinant): the determinant associated to the standard basis of the free module is the usual matrix determinant

ocfnash opened the pull request:
Formalized as part of the Sphere Eversion project.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(linear_algebra/determinant): `matrix.det` as an `alternating_map` on the free module

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -400,3 +400,12 @@ by rw [basis.det_reindex, function.comp.assoc, e.self_comp_symm, function.comp.r
 by { rw [basis.det_apply, basis.to_matrix_map, basis.det_apply] }
+
+/-- This lemma is useful because it enables us to see `matrix.det` as (the coercion of) an
+`alternating_map` on the module `ι → R`. -/
```
Is that different to what `matrix.det_row_multilinear` is?

eric-wieser replied:
Maybe simply replacing
```lean
abbreviation det (M : matrix n n R) : R :=
det_row_multilinear M
```
with
```lean
abbreviation det : matrix n n R → R :=
(det_row_multilinear : alternating_map R _ _ n)
```
achieves what you need here, then `dunfold matrix.det` introduces such an alternating map without an annoying lambda.

ocfnash replied:
Oh, thanks. How did I miss `matrix.det_row_multilinear`; it should indeed be sufficient for my purposes.

I'll come back to this in a few days once I've finished the other parts of what I need.

eric-wieser replied:
Perhaps it should be renamed `det_row_alternating`; I didn't change the name when I upgraded it to an alternating map.

I think the lemmas in this PR are still worth having, maybe just this one in the other direction and without the misleading docstring motivation

ocfnash replied:
Agreed on all points. I have attempted to implement these suggestions.

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -400,3 +400,12 @@ by rw [basis.det_reindex, function.comp.assoc, e.self_comp_symm, function.comp.r
+  change _ = matrix.det ∘ (pi.basis_fun R ι).to_matrix,
+  ext M,
+  simp [basis.coe_pi_basis_fun.to_matrix_eq_transpose],
+end
```
golfed:
```suggestion
begin
  ext M,
  rw [basis.det_apply, basis.coe_pi_basis_fun.to_matrix_eq_transpose, det_transpose],
end
```
I'd be tempted to reverse the direction of the statement since the RHS looks more complex.

ocfnash replied:
Thanks, done.

ocfnash committed:
CR suggestions: flip lemma direction, rename, improve proof, add `simp`, drop comment

ocfnash committed:
Suggestions of CR: rename `matrix.det_row_multilinear` to `matrix.det_row_alternating`

This is not directly connected with the other changes in this PR but this seems like
a good time to make this change.

ocfnash committed:
Upgrade lemma

jcommelin reviewed (approved):
Thanks :tada:

bors merge

ocfnash committed:
merge master
